de-lib/src/main/java/org/iplantc/de/diskResource/client/views/metadata/dialogs/MetadataTemplateViewDialog.java

```diff
@@ -100,45 +111,63 @@ public MetadataTemplateViewDialog(MetadataView.Presenter presenter,List<Avu> tem
 
                 avu.setValue(value);
             }
-
             avus.add(avu);
-        }
+
+        });
         return avus;
     }
-    
+
     public boolean isValid() {
-      List<IsField<?>> fields = FormPanelHelper.getFields(widget);
-            for (IsField<?> f : fields) {
-                if (!f.isValid(false)) {
-                    valid = false;
-                }
+        List<IsField<?>> fields = FormPanelHelper.getFields(widget);
+        for (IsField<?> f : fields) {
+            if (!f.isValid(false)) {
+                valid = false;
             }
+        }
 
-      return valid;
+        return valid;
     }
-    
+
     private void buildAvuMap() {
-    	for (MetadataTemplateAttribute attribute: attributes) {
-    		Avu md = findMetadataForAttribute(attribute.getName());
-    		templateAttrAvuMap.put(attribute.getName(),md);
-    	}
+        for (MetadataTemplateAttribute attribute : attributes) {
+            List<Avu> mds = findMetadataForAttribute(attribute.getName());
+            if (mds != null && !mds.isEmpty()) {
+                mds.forEach(md -> {
+                    String tag = getTagForMetadata(md);
+                    templateTagAvuMap.put(tag, md);
+                    templateTagAtrrMap.put(tag, attribute);
+                    // GWT.log(tag + "-->" + attribute.getName());
+                });
+            } else {
+                Avu avu = presenter.setAvuModelKey(MetadataPresenterImpl.newMetadata(attribute.getName(),
+                                                                                     "",
+                                                                                     "")); //$NON-NLS-1$ //$NON-NLS-2$
+                String tag = getTagForMetadata(avu);
+                templateTagAvuMap.put(tag, avu);
+                templateTagAtrrMap.put(tag, attribute);
+                // GWT.log(tag + "-->" + attribute.getName());
+            }
+
+        }
     }
-    
-    private Avu findMetadataForAttribute(String attribute) {
-    	if(templateMd != null) {
-	    	for(Avu md: templateMd) {
-	    		if(md.getAttribute().equals(attribute)) {
-	    			return md;
-	    		}
-	    	}
-    	}
-		return null;
+
+    private List<Avu> findMetadataForAttribute(String attribute) {
```

Comment:
I will summarize my response here for above comments about moving some code to Presenter.  I feel they should stay here because all these logic is about managing / manipulating the template view. The `FastMap`s in this class is just to manage how attributes, fields and values are linked. Also, all these values does not get saved to the server from here. Rather they get loaded into the grid view through the presenter.